### DISCUSSION TITLE
nested computed getters unexpected

### DISCUSSION THREAD

**Discussion by magicdawn** [pmndrs/valtio#692] [General]
CSB
https://codesandbox.io/s/valtio-countdown-to-do-list-forked-c4bii0?file=/src/App.tsx

valtio has deprecated `proxyWithComputed` and recommend use `getter` to implement computed.
I never read the source code of `proxyWithComputed`, but a lot with getter computed because I'm troubled.

## `useSnapshot`

normal case if our code is using `useSnapshot`. 

- after any field change 
- decide whether snapshot has changed
- call `snapshot()`
- this will run all `get xxxgetter(){ }`

getters are evaluated frequently, so we need cache.

## cache is hard

recommend way

```ts
import {memoize} from 'proxy-memoize'

const _c = memoize(snap => snap.a + snap.b)

const store = proxy({
  a: 1,
  b: 2,
  get c() {
    return _c(snapshot(store))
  }
})
```

this pattern has unexpected when computed nested

```ts
const _x = memoize((snap: any) => snap.y * snap.y);
const _y = memoize((snap: any) => snap.a + snap.b);
const _z = memoize((snap: any) => snap.y * snap.y); // _x same as _z

const store = proxy({
  a: 1,
  b: 2,
  get x() {
    console.log("calculating x");
    return _x(snapshot(store));
  },
  get y() {
    return _y(snapshot(store));
  },
  get z() {
    return _z(snapshot(store));
  }
});
```

for `x`, inside `get x(){}` `snapshot(store).y` is `undefined` because inside `snapshot()` use `Reflect.ownKeys(target).forEach((key) =>{})`, `x` index smaller than `y`

for `z`, inside `get z(){}` `snapshot(store).y` has value, but still NaN in `snap.y`

https://codesandbox.io/s/valtio-countdown-to-do-list-forked-c4bii0?file=/src/App.tsx

with valtio@1.5.2
![image](https://user-images.githubusercontent.com/4067115/223309321-85672c90-eb3f-419b-8c88-239f38896bcc.png)

with valtio@1.10.3
![image](https://user-images.githubusercontent.com/4067115/223309584-f0c9023d-6ede-43f7-816e-9496f23a2736.png)

I have tried remove `snapshot()` in getters and manualy select dep like `store.y`, make snapshot(dep), and use plain `memoize-one`, but `snapshot()` is not smart enough.
if store.y returns a array, in  x or z, store.y switchs between `Proxy target an Array` and `An array of Proxied Items`. `snapshot()` fails on `An array of Proxied Items`.

SO I end up with no cache. Can't make things work with cache 😂

**Comment by dai-shi** [comment]
Sorry for the confusion.
You have to pass a snapshot object to a memoized function.

```diff
  import {memoize} from 'proxy-memoize'
  
  const _c = memoize(snap => snap.a + snap.b)
  
  const store = proxy({
    a: 1,
    b: 2,
    get c() {
-     return _c(snapshot(store))
+     return _c(snapshot(snapshot(store)))
    }
  })
```

**Reply by dai-shi** [reply]
My bad totally. I was sleeping probably.

**Comment by magicdawn** [comment]
looks like the double snapshot not working

![image](https://user-images.githubusercontent.com/4067115/224120596-9ddbf35c-9980-4d63-a4e7-c6e5eaec38b9.png)

**Reply by dai-shi** [reply]
Wow, what did I do? My ignorance. Please forget about it...

**Comment by dai-shi** [comment]
Re-reading the description, I think nested computeds are not supported now.
Does it work with deprecated `proxyWithComputed`?

Maybe a workaround is to use `derive` instead, but it may not be very handy.

**Reply by magicdawn** [reply]
proxyWithComputed can not be used as nested 

```
(alias) function proxyWithComputed<T extends object, U extends object>(initialObject: T, computedFns: { [K in keyof U]: ((snap: Snapshot<T>) => U[K]) | {
    get: (snap: Snapshot<T>) => U[K];
    set?: ((state: T, newValue: U[K]) => void) | undefined;
}; }): T & U
```
snap is Snapshot<T> ...

--------

AND derive works.

Thanks for reply.

**Comment by ssijak** [comment]
can it be written like this @dai-shi ?
```
import {memoize} from 'proxy-memoize'

const store = proxy({
  a: 1,
  b: 2,
  get c() {
    const _c = memoize(snap => snap.a + snap.b)
    return _c(snapshot(store))
  }
})
```

or `const _c = memoize(snap => snap.a + snap.b)` needs to be outside of the getter?

**Reply by dai-shi** [reply]
yeah, memoize should be outside of the getter.
